NEW Ensure polymorphic has_one fields are indexed

robbieaverill opened the pull request:
I'm not sure whether this is the best way to tackle this, but at the point where you're looping database fields there's nothing to indicate that the relationship is polymorphic other than it is a ClassName enum. I've used that to look for a matching field name with the "ID" suffix and treat those as polymorphic.

Resolves #6966

tractorcow commented on src/ORM/DataObjectSchema.php:
```diff
@@ -486,29 +486,76 @@ protected function cacheDatabaseFields($class)
+    {
+        $indexes = [];
+        $databaseFields = $this->databaseFields($class, false);
+        foreach ($databaseFields as $field => $type) {
```
Would it be more efficient to simply loop over all has_one? At least you would have exactly one index per loop. Then you only need to compare the right-hand side of the has_one to check if it's polymorphic or not. I'd abandon the old implementation and start afresh.

dhensby commented:
In my dbal work I did the following:

```php
                /** @var DBField $field */
                $fieldSpec = ClassInfo::parse_class_spec($fieldType);
                array_unshift($fieldSpec[1], $fieldName);
                $field = Injector::inst()->createWithArgs($fieldSpec[0], $fieldSpec[1]);
                $field->augmentDBTable($dbTable);
```

You could do this and then perform an `instanceof` or, better yet, let the field type itself determine if there's an index to add... `$field->augmentDBIndex($indexes);`

tractorcow commented:
I was going to say parse_class_specs might be a bit inefficient for has_ones, but I guess this is only done on schema update right (dev/build).

@dhensby what was your outer loop for that code?

robbieaverill commented:
@tractorcow @dhensby I've pushed a different approach which hopefully is closer to what you've both suggested.

Couple of notes:

* I've added some logic to track polymorphic indexes in the `cacheDatabaseFields` method, since that is the point where PolymorphicForeignKey instances are converted to Int and ClassName and are difficult to detect afterwards
* Have made "indexed" configurable
* Still need to loop `databaseFields()` as we need to add indexes on some fixed fields like ClassName which wouldn't be picked up if we simply looped the has_ones

dhensby commented:
I'm a little unclear on this because I'd expect that we use the `composite_db` config on composite fields to build the index but I don't see that happening anywhere?

tractorcow commented:
Well, not every composite field is automatically an index in the same way that has_ones are.

tractorcow commented on src/ORM/DataObjectSchema.php:
```diff
@@ -486,29 +500,65 @@ protected function cacheDatabaseFields($class)
+     * @param  string $class
+     * @return array
+     */
+    protected function defaultDatabaseIndexes($class)
```
Should be called `cacheDefaultDatabaseIndexes` to distinguish this from the public getter and keep its purpose clear. This should also be the only place that defaultDatabaseIndexes is populated, since it seems part of this method has migrated to cacheDatabaseFields.

tractorcow commented on src/ORM/DataObjectSchema.php:
```diff
@@ -486,29 +500,65 @@ protected function cacheDatabaseFields($class)
+     * @throws InvalidArgumentException If an index already exists on the class
+     * @throws InvalidArgumentException If a custom index format is not valid
+     */
+    protected function customDatabaseIndexes($class)
```
As with `buildTableName` being called from `cacheTableNames`, this should be called `buildCustomDatabaseIndexes` as it's called from `cacheDatabaseIndexes`, so that it doesn't look like the standard getter.

tractorcow commented on src/ORM/DataObjectSchema.php:
```diff
@@ -471,7 +479,13 @@ protected function cacheDatabaseFields($class)
+                $compositeName = $fieldName . $nestedName;
+                $dbFields[$compositeName] = $nestedSpec;
+
+                // Handle configurable indexes, e.g. for polymorphic keys
```
This logic should probably be in defaultDatabaseIndexes shouldn't it?

robbieaverill replied:
Yes ideally, but at that point you don't know whether it's a polymorphic key because at this point it gets changed into an Int and a DBClassName instead

tractorcow replied:
My suggestion for using fieldSpecs should pull those out in the result. Although, you'll get both the composite field parent, and all sub-fields in the result. Not an issue unless a composite field has 'indexed' and one of it's components ALSO has 'indexed' though right?

tractorcow commented on src/ORM/DataObjectSchema.php:
```diff
@@ -486,29 +500,65 @@ protected function cacheDatabaseFields($class)
-                    'columns' => [$field],
-                ];
+        // Ensure that fixed field types, e.g. ClassName can be indexed
+        $databaseFields = $this->databaseFields($class, false);
```
If you use `$this->fieldSpecs($class, UNINHERITED)` here you would get all DBField instances, including DBPolymorphicForeignKey, DBForeignKey, and other instances.

All you need is the class of the field, and the `indexed` config option for that class right? then all you need is a method that gets the list of fields to index for a given DBField instance.

robbieaverill replied:
Actually, I'm not sure we can do this just with config - `DBPolymorphicForeignKey` is not a FQCN. We could define unqualified class name keys in a config file somewhere but then they wouldn't be bound to a concrete class. Dynamic options like `$className::class` don't work.

robbieaverill replied:
Oh, this can be looked up from injector config

robbieaverill replied:
No - we need a class instance to check if it's a composite field or not so composite fields can have their own fields indexed

tractorcow commented on src/ORM/DataObjectSchema.php:
```diff
@@ -486,29 +500,65 @@ protected function cacheDatabaseFields($class)
 
-        // look for custom indexes declared on the class
+        // Return the registered default indexes
+        foreach ($this->defaultDatabaseIndexes[$class] as $field) {
```
What if $field is a composite field? `colums` should be as many fields as that composite field has right?

tractorcow commented:
This needs some tests with polymorphic has_ones, normal composite fields, and maybe DBClassName for good measure, to make sure the appropriate indexes are built. :) I've left all feedback I could!

Good job so far.

robbieaverill commented:
Thanks, will add some more tests when I get a chance

dhensby commented:
> Well, not every composite field is automatically an index in the same way that has_ones are.

Why is this? If it's not then why are we building this into core :/ - I'm probably missing something fairly obvious here, sorry.

I'd have thought that the composite field ID/ClassName should always be indexed

tractorcow commented:
Yes, but composite Money Field (currency, amount) shouldn't automatically be indexed.

dhensby commented:
ah - ok

robbieaverill commented:
Ok, I've rebased and pushed a new commit that will allow composite fields to be indexed with their fields together, and treats polymorphic has_ones as a composite field in that regard.

I've updated some of the bits as per @tractorcow's suggestions.

Need to add tests another day.

robbieaverill commented:
Ok, ready for re-review

tractorcow commented on src/ORM/DataObjectSchema.php:
```diff
@@ -486,30 +494,72 @@ protected function cacheDatabaseFields($class)
+            $fieldObj = Injector::inst()->create($fieldClass, $field);
+
+            // Polymorphic indexes are handled as a DBComposite, check to avoid adding them twice
+            if ($fieldObj->config()->get('indexed') && array_shift($arguments) !== DataObject::class) {
```
What was that dataobject class check for? Is that `DBClassName(DataObject)`?

robbieaverill replied:
It's for catching polymorphics that have a class name added in - we don't need to index `PolymorphicClassName` on its own as well as the combination of PolymorphicID and *ClassName

tractorcow replied:
Oh, I see. In that case, I suggest moving field-specific index generation (such as this) to the DBPolymorphicForeignKey instance. See my other suggestion.

tractorcow replied:
Hm, maybe some check that skips this field if it's a member of a composite field?

tractorcow replied:
Suggest to control it via spec arguments directly.

```php
```private static $composite_db = array(
        'ID' => 'Int',
        'Class' => "DBClassName('SilverStripe\\ORM\\DataObject', [ 'indexed' => false ])"
    );
```

Similarly to how we specify options for HTMLText.

tractorcow commented on src/ORM/DataObjectSchema.php:
```diff
@@ -486,30 +494,72 @@ protected function cacheDatabaseFields($class)
+            // Polymorphic indexes are handled as a DBComposite, check to avoid adding them twice
+            if ($fieldObj->config()->get('indexed') && array_shift($arguments) !== DataObject::class) {
+                // Either index a composite field's fields, or just the field itself
+                if ($fieldObj instanceof DBComposite) {
```
Could we move this logic into DBField itself? It might be useful for DBFields subclasses to customise their own index specs.

robbieaverill commented:
Action points:

* Remove polymorphic check from DataObjectSchema
* Add constructor arg to DBField for setting bool indexed with setter/getter
* Constructor arg or `setIndexed()` value takes precendence over global config for DBField type
* Move logic for "get fields to include in index" to DBField and implement in DBComposite to use `composite_db`, default value in DBField would be the field name

tractorcow commented:
> Add constructor arg to DBField for setting bool indexed with setter/getter

Be careful, since there's already an $options in DBString. Should that just go up one level to DBField?

robbieaverill commented:
>Be careful, since there's already an $options in DBString. Should that just go up one level to DBField?

I think that makes sense

robbieaverill commented:
Ok, good to go again. This isn't really a patch fix any more - I can squash and remove "FIX" from one of the commits when we're good to go!

tractorcow commented:
It's looking really good @robbieaverill . I'll do some more peer review later this week when I check it out and test it.

tractorcow commented on src/ORM/DataObjectSchema.php:
```diff
@@ -486,30 +493,59 @@ protected function cacheDatabaseFields($class)
+            $fieldObj = Injector::inst()->create($spec, $field);
+            if ($fieldObj->getIndexed()) {
+                $this->defaultDatabaseIndexes[$class][$field] = [
                     'type' => 'index',
```
If you move the whole array syntax to the DBField, theoretically you could have field-specific index types (e.g. unique, index, fulltext).

robbieaverill replied:
Good call - updated in Good call - updated in 345357f which integrates this into the original approach to keep it simple yet more flexible

dhensby commented:
@tractorcow should this make the 4.0 release or 4.1? If 4.0 then it needs to go into the 4.0 branch

dhensby commented:
Nice - this is starting to look a lot like my DBAL approach ;)

robbieaverill commented:
Happy to rebase and squash as appropriate when requested. Will leave the commits as is for now to help with following peer review

robbieaverill commented:
Rebased

tractorcow commented:
@robbieaverill where is 345357f?

robbieaverill commented:
@tractorcow - sorry, it got killed in the rebase

tractorcow commented:
@robbie what I was suggesting was
```
$this->defaultDatabaseIndexes[$class][$field] = fieldObj->getIndexSpecs();
```

not

```
$this->defaultDatabaseIndexes[$class][$field] = [
    'type' => $type,
    'columns' => $fieldObj->getIndexSpecs(),
];
```

robbieaverill commented:
@tractorcow thanks for the review. Implementors of the `DBIndexable` interface are now responsible for returning a full and complete index spec (as validated by `DBSchemaManager::convertIndexSpec`).
